api-fput-object.go

```diff
@@ -139,49 +141,100 @@ func (c Client) FPutObject(bucketName, objectName, filePath, contentType string)
 		return 0, err
 	}
 
-	// gather next part number to be uploaded, total uploaded size and list of all parts uploaded.
-	partNumber, totalUploadedSize, completeMultipartUpload, err := c.getMultipartStat(bucketName, objectName, uploadID)
+	// total data read and written to server. should be equal to 'size' at the end of the call.
+	var totalUploadedSize int64
+
+	// Complete multipart upload.
+	var completeMultipartUpload completeMultipartUpload
+
+	// Fetch previously upload parts and save the total size.
+	partsInfo, err := c.listObjectParts(bucketName, objectName, uploadID)
 	if err != nil {
 		return 0, err
 	}
+	// Save uploaded parts in a map.
+	var uploadedPartsMap = make(map[int]completePart)
+	// Previous maximum part size
+	var prevMaxPartSize int64
+	// Loop through all parts to save them in a local map.
+	for _, partInfo := range partsInfo {
+		var completedPart completePart
+		completedPart.ETag = partInfo.ETag
+		completedPart.PartNumber = partInfo.PartNumber
+		uploadedPartsMap[partInfo.PartNumber] = completedPart
+		totalUploadedSize += partInfo.Size
+		if partInfo.Size >= prevMaxPartSize {
+			prevMaxPartSize = partInfo.Size
+		}
+	}
 
 	// Calculate the optimal part size for a given file size.
-	partSize := calculatePartSize(fileSize)
-
-	// Seek to the total uploaded size obtained after listing all the parts.
-	if totalUploadedSize > 0 {
-		if _, err := fileData.Seek(totalUploadedSize, 0); err != nil {
-			return 0, err
-		}
+	partSize := optimalPartSize(fileSize)
+	// If prevMaxPartSize is set use that.
+	if prevMaxPartSize != 0 {
+		partSize = prevMaxPartSize
 	}
 
-	// Make done channel.
-	doneCh := make(chan struct{})
-	// Close done channel to indicate closure to all go-routines.
-	defer close(doneCh)
+	// MD5 and Sha256 hasher.
+	var hashMD5, hashSha256 hash.Hash
+
+	// Part number always starts with '1'.
+	partNumber := 1
 
-	// Loop through all sections of the file.
-	for part := range sectionManager(fileData, fileSize, partSize, enableSha256Sum, doneCh) {
-		// On any error return.
-		if part.Err != nil {
-			return 0, part.Err
+	// Loop through until EOF.
+	for totalUploadedSize < fileSize {
+		// Create a hash multiwriter.
+		hashMD5 = md5.New()
+		hashWriter := io.MultiWriter(hashMD5)
+		if c.signature.isV4() {
+			hashSha256 = sha256.New()
```

Comment:
This can be turned into a function to compute hash without disturbing the seek offset.